### PR TITLE
Caprevoke packages

### PR DESCRIPTION
**Pull request by nwf-msr**
After much delay and yak shaving, here is a first attempt at adding packages for
- `snmalloc` shims, both upstream (which {c,sh}ould probably go upstream to FreeBSD?) and with extensions for mrs
- a `jemalloc`-for-`mrs` shim
- `mrs` shims configured as per CHERIvoke, Cornucopia, Cornucopia-successor (load barriers), and Cornucopia-successor-but-with-statistics.

I'll get instructions for how to pull the right kernels and then to use these up in the cheribsd wiki.  In the interim, it'd be really good to have several more competent people look these over as I definitely felt like I was cutting against the grain writing these (I guess master/subordinate packages are not exactly the common case, though).

### CODE REVIEW

File: devel/snmalloc/Makefile [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+MAINTAINER=	nfilardo@microsoft.com
+COMMENT=	High-performance memory allocator
+WWW=		https://github.com/microsoft/snmalloc
+
+LICENSE=	MIT
+LICENSE_FILE=	${WRKSRC}/LICENSE
+
+USES=		cmake:testing compiler:c++17-lang
+
+PLIST_SUB=	_libdirsfx=${PKGNAMESUFFIX}
```
Comment (brooksdavis) [inline devel/snmalloc/Makefile:14]: It's more common to just use `PKGNAMESUFFIX=${PKGNAMESUFFIX}` and `%% PKGNAMESUFFIX%%` in the plist.

File: devel/mrs/Makefile [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ALL_TARGET?=	mrs
+
+.if ${PKGNAMESUFFIX:U} == ""
+CMAKE_ON+=	OFFLOAD_QUARANTINE LOAD_SIDE_REVOCATION
+CMAKE_ARGS+=	-DCONCURRENT_REVOCATION_PASSES=1 \
+		-DQUARANTINE_RATIO=4
+.endif
+
+.if !target(do-install)
+PLIST_SUB=	_libsfx=${PKGNAMESUFFIX}
```
Comment (brooksdavis) [inline devel/mrs/Makefile:32]: This stomps on the PLIST_SUB assignment above.

File: devel/jemalloc-mrs/pkg-descr [hunk 3]
```diff
@@ -0,0 +1,5 @@
+
```
Comment (brooksdavis) [inline devel/jemalloc-mrs/pkg-descr:1]: To pick on a arbitrary instance. I don't think it's common style to have blank lines at the top of pkg-desc.

**Review by brooksdavis — COMMENTED** [review]
Generally looks good.

I'd be somewhat tempted to make the various mrs-* port into flavors using the FLAVOR mechanism. jemalloc-mrs somewhat complicates this idea.

All the ports might benefit from pkg-message files that provide LD_PRELOAD examples.

**Comment by nwf-msr** [comment]
> Generally looks good.

Huzzah.

> I'd be somewhat tempted to make the various mrs-* port into flavors using the FLAVOR mechanism. jemalloc-mrs somewhat complicates this idea.

Maybe it's worth having `jemalloc-mrs` stand alone and not depend on `mrs`.  I'll take a stab at `FLAVOR`ing.

> All the ports might benefit from pkg-message files that provide LD_PRELOAD examples.

Can do.

**Comment by nwf** [comment]
Thanks for the suggestions.  FLAVORS turned out to be much less of a pain than I'd feared, I think.

**Comment by nwf** [comment]
Hold fire; testing has revealed some small bugs (as should be expected).

File: devel/mrs/Makefile [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+USE_GITHUB=	yes
+GH_ACCOUNT=	nwf-msr
+GH_PROJECT=	mrs
+GH_TAGNAME=	d77220664c077d0568019eeb5f4d1d1dacea5343
+
+ALL_TARGET=	mrs
+
+CMAKE_ARGS+=	-DGIT_SHA1_OVERRIDE=${GH_TAGNAME}
+
+.if ${FLAVOR:M*_ratio*} == ""
```
Comment (brooksdavis) [inline devel/mrs/Makefile:32]:
```suggestion
.if !${FLAVOR:M*_ratio*}
```
I tend to prefer using :M as a boolean operator in these kinds of things.

File: devel/mrs/Makefile [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+CMAKE_ARGS+=	-DQUARANTINE_RATIO=4
+.endif
+
+.if ${FLAVOR:Mcherivoke*} != ""
```
Comment (brooksdavis) [inline devel/mrs/Makefile:36]:
```suggestion
.if ${FLAVOR:Mcherivoke*}
```

**Review by brooksdavis — APPROVED** [review]
I like this.  FLAVORs really improved the mrs port. Much more comprehensible.

**Comment by nwf-msr** [comment]
Things look better now; SWIM want to give this another once-over or try following the instructions, such as they are, at https://github.com/CTSRD-CHERI/cheripedia/wiki/HOWTO:-Use-Cornucopia-with-the-22.12-CheriBSD-Release ?

**Comment by kwitaszczyk** [comment]
Should these ports only be available for aarch64c? Or would they work for aarch64, riscv64c, riscv64 as well?

**Comment by kwitaszczyk** [comment]
The ports also must be added to devel/Makefile to be built with poudriere-bulk.

**Review by kwitaszczyk — APPROVED** [review]

**Pull request merged**